cibyl/outputs/cli/ci/system/impls/zuul/colored.py

```diff
@@ -44,80 +45,78 @@ class ProjectCascade(ColoredPrinter):
+                        result.add(self._print_pipeline(project, pipeline), 1)
+
+                    msg = "Total pipelines found in query for project '"
+                    result.add(self.palette.blue(msg), 1)
+                    result[-1].append(self.palette.underline(project.name))
+                    result[-1].append(self.palette.blue("': "))
+                    result[-1].append(len(project.pipelines))
+                else:
+                    msg = 'No pipelines found in query.'
+                    result.add(self.palette.red(msg), 1)
 
             return result.build()
 
         def _print_pipeline(self, project, pipeline):
             result = IndentedTextBuilder()
 
-            result.add(self._palette.blue('Pipeline: '), 0)
+            result.add(self.palette.blue('Pipeline: '), 0)
             result[-1].append(pipeline.name)
 
             if self.query >= QueryType.JOBS:
-                for job in pipeline.jobs.values():
-                    result.add(self._print_job(project, pipeline, job), 1)
-
-                result.add(
-                    self._palette.blue(
-                        "Total jobs found in query for pipeline '"
-                    ), 1
-                )
-
-                result[-1].append(self._palette.underline(pipeline.name))
-                result[-1].append(self._palette.blue("': "))
-                result[-1].append(len(pipeline.jobs))
+                if pipeline.jobs.value:
+                    for job in pipeline.jobs.values():
+                        result.add(self._print_job(project, pipeline, job), 1)
+
+                    msg = "Total jobs found in query for pipeline '"
+                    result.add(self.palette.blue(msg), 1)
+                    result[-1].append(self.palette.underline(pipeline.name))
+                    result[-1].append(self.palette.blue("': "))
+                    result[-1].append(len(pipeline.jobs))
+                else:
+                    msg = 'No jobs found in query.'
+                    result.add(self.palette.red(msg), 1)
 
             return result.build()
 
         def _print_job(self, project, pipeline, job):
             result = IndentedTextBuilder()
 
-            result.add(self._palette.blue('Job: '), 0)
+            result.add(self.palette.blue('Job: '), 0)
             result[-1].append(job.name.value)
 
             if self.query >= QueryType.BUILDS:
-                for build in job.builds.values():
-                    # This cascade only wants builds of this project
-                    if build.project.value != project.name.value:
-                        continue
-
-                    # This cascade only wants builds triggered by this pipeline
-                    if build.pipeline.value != pipeline.name.value:
-                        continue
+                builds = apply_filters(
+                    job.builds.values(),
+                    lambda bld: bld.project.value != project.name.value,
```

Comment:
Why filtering the builds here? If it's a temporary solution for some issue with zuul I believe we should have a comment explaining the reason. Long term we should avoid having such logic in the printer.

Reply:
Because for that particular case, out of all the builds the job has, I only want to print the ones under a certain project and pipeline. This is not a temporary solution, it is a condition that the printer has to meet.

Reply:
If the user passes some values for projects and pipelines, shouldn't the zuul source already filter the build so that only does under those projects and pipelines are added?

Reply:
And it does, the resulting builds are all stored under their job. A job can hold builds for multiple projects and pipelines, so what I do here is recover the builds that were found for a particular combination.

All builds under a job are printed on the JobCascade class. It is the same thing, but with two different representations.

Reply:
All right, makes sense then, thanks for the clarification